src/interface/swarm.cpp

```diff
@@ -317,28 +354,537 @@ void Swarm::Defrag() {
+          } else if (k == 0) {
+            neighborIndices_h(1, 3, 0) = n;
+            neighborIndices_h(2, 3, 0) = n;
+          } else if (k == 1) {
+            neighborIndices_h(3, 3, 0) = n;
+          }
+        }
+      } else if (i == 0) {
+        if (j == -1) {
+          if (k == -1) {
+            neighborIndices_h(0, 0, 1) = n;
+            neighborIndices_h(0, 0, 2) = n;
+          } else if (k == 0) {
+            neighborIndices_h(1, 0, 1) = n;
+            neighborIndices_h(1, 0, 2) = n;
+            neighborIndices_h(2, 0, 1) = n;
+            neighborIndices_h(2, 0, 2) = n;
+          } else if (k == 1) {
+            neighborIndices_h(3, 0, 1) = n;
+            neighborIndices_h(3, 0, 2) = n;
+          }
+        } else if (j == 0) {
+          if (k == -1) {
+            neighborIndices_h(0, 1, 1) = n;
+            neighborIndices_h(0, 1, 2) = n;
+            neighborIndices_h(0, 2, 1) = n;
+            neighborIndices_h(0, 2, 2) = n;
+          } else if (k == 1) {
+            neighborIndices_h(3, 1, 1) = n;
+            neighborIndices_h(3, 1, 2) = n;
+            neighborIndices_h(3, 2, 1) = n;
+            neighborIndices_h(3, 2, 2) = n;
+          }
+        } else if (j == 1) {
+          if (k == -1) {
+            neighborIndices_h(0, 3, 1) = n;
+            neighborIndices_h(0, 3, 2) = n;
+          } else if (k == 0) {
+            neighborIndices_h(1, 3, 1) = n;
+            neighborIndices_h(1, 3, 2) = n;
+            neighborIndices_h(2, 3, 1) = n;
+            neighborIndices_h(2, 3, 2) = n;
+          } else if (k == 1) {
+            neighborIndices_h(3, 3, 1) = n;
+            neighborIndices_h(3, 3, 2) = n;
+          }
+        }
+      } else if (i == 1) {
+        if (j == -1) {
+          if (k == -1) {
+            neighborIndices_h(0, 0, 3) = n;
+          } else if (k == 0) {
+            neighborIndices_h(1, 0, 3) = n;
+            neighborIndices_h(2, 0, 3) = n;
+          } else if (k == 1) {
+            neighborIndices_h(3, 0, 3) = n;
+          }
+        } else if (j == 0) {
+          if (k == -1) {
+            neighborIndices_h(0, 1, 3) = n;
+            neighborIndices_h(0, 2, 3) = n;
+          } else if (k == 0) {
+            neighborIndices_h(1, 1, 3) = n;
+            neighborIndices_h(1, 2, 3) = n;
+            neighborIndices_h(2, 1, 3) = n;
+            neighborIndices_h(2, 2, 3) = n;
+          } else if (k == 1) {
+            neighborIndices_h(3, 1, 3) = n;
+            neighborIndices_h(3, 2, 3) = n;
+          }
+        } else if (j == 1) {
+          if (k == -1) {
+            neighborIndices_h(0, 3, 3) = n;
+          } else if (k == 0) {
+            neighborIndices_h(1, 3, 3) = n;
+            neighborIndices_h(2, 3, 3) = n;
+          } else if (k == 1) {
+            neighborIndices_h(3, 3, 3) = n;
+          }
+        }
+      }
+    }
+  }
+
+  neighborIndices_.DeepCopy(neighborIndices_h);
+}
+
+bool Swarm::Send(BoundaryCommSubset phase) {
```

Comment:
This method is too long, I would also recommend breaking into subfunctions for better readability. There are several areas I think would be appropriate e.g. the for loops.

Reply:
I'm trying to follow the logic here, but could use a little help, why can this method not be made constant, when I think of sending information my intuition tells me that I should not be changing the state of the Swarm, I see that you are changing vbswarm, but that is a shared pointer which indicates the Swarm does not have ownership anyway so making it constant should still be ok. But what am I missing. 

I would prefer to see logic like a method call PreparePackage/PrepareSwarm which is non constant followed by a constant Send. This might be wishful thinking on my part though.

Reply:
Another option might be to change the method name to PrepareAndSend or something like that.

Reply:
OK I broke this up into descriptive subfunctions. Does this flow better? It also turns out the physical boundaries application was unnecessary (this is already captured in all cases of [mpi send/same mpi rank copy] by the receiving meshblock) so I was able to remove a good amount of potentially expensive code.

I'm hesitant to break this into more than one call because that would mean adding extra tasks for the app that *must* all be included in the correct order. I'm also not sold yet on expanding the name of this function since the app dev doesn't need to know anything about how we load/unload buffers as part of boundary communication, they just need to tell the code to send and receive boundaries and the rest is handled behind the scenes.

Reply:
Ok.